secrethub_wrapper.go

```diff
@@ -0,0 +1,187 @@
+			LatestVersion: C.int(secret.Secret.LatestVersion),
+			Status:        C.CString(secret.Secret.Status),
+			CreatedAt:     C.longlong(secret.Secret.CreatedAt.Unix()),
+		},
+		Version:   C.int(secret.Version),
+		Data:      C.CString(string(secret.Data)),
+		CreatedAt: C.longlong(secret.CreatedAt.Unix()),
+		Status:    C.CString(secret.Status),
+	}
+}
+
+// ReadString retrieves a secret as a string.
+//export ReadString
+func ReadString(path *C.char, errMessage **C.char) *C.char {
+	client, err := Client()
+	if err != nil {
+		*errMessage = C.CString(err.Error())
+		return nil
+	}
+	secret, err := client.Secrets().Read(C.GoString(path))
+	if err != nil {
+		*errMessage = C.CString(err.Error())
+		return nil
+	}
+	return C.CString(string(secret.Data))
+}
+
+// Resolve fetches the values of a secret from SecretHub, when the `ref` parameter
+// has the format `secrethub://<path>`. Otherwise it returns `ref` unchanged, as an array of bytes.
+//export Resolve
+func Resolve(ref *C.char, errMessage **C.char) *C.char {
+	client, err := Client()
+	if err != nil {
+		*errMessage = C.CString(err.Error())
+		return nil
+	}
+	bits := strings.Split(C.GoString(ref), "://")
+	if len(bits) == 2 && bits[0] == "secrethub" {
+		secret, err := client.Secrets().Read(bits[1])
+		if err != nil {
+			*errMessage = C.CString(err.Error())
+			return nil
+		}
+		return C.CString(string(secret.Data))
+	}
+	return ref
+}
+
+// ResolveEnv takes a map of environment variables and replaces the values of those
+// which store references of secrets in SecretHub (`secrethub://<path>`) with the value
+// of the respective secret. The other entries in the map remain untouched.
+//export ResolveEnv
+func ResolveEnv(errMessage **C.char) *C.char {
+	envVars := os.Environ()
+	resolvedEnv := make(map[string]string, len(envVars))
+	for _, value := range envVars {
+		envVar := strings.Split(value, "=")
+		key := envVar[0]
+		value := C.CString(envVar[1])
+		resolvedValue := Resolve(value, errMessage)
+		resolvedEnv[key] = C.GoString(resolvedValue)
+	}
+	encoding, err := json.Marshal(resolvedEnv)
+	if err != nil {
+		*errMessage = C.CString(err.Error())
+		return nil
+	}
+	return C.CString(string(encoding))
+}
+
+// Exists checks if a secret exists at `path`.
+//export Exists
+func Exists(path *C.char, errMessage **C.char) C.bool {
+	client, err := Client()
+	if err != nil {
+		*errMessage = C.CString(err.Error())
+		return C.bool(false)
+	}
+	exists, err := client.Secrets().Exists(C.GoString(path))
+	if err != nil {
+		*errMessage = C.CString(err.Error())
+		return C.bool(false)
+	}
+	return C.bool(exists)
+}
+
+// Remove deletes the secret found at `path`, if it exists.
+//export Remove
+func Remove(path *C.char, errMessage **C.char) {
+	client, err := Client()
+	if err != nil {
+		*errMessage = C.CString(err.Error())
+		return
+	}
+	err = client.Secrets().Delete(C.GoString(path))
+	if err != nil {
+		*errMessage = C.CString(err.Error())
```

Comment:
Though technically not necessary, to avoid future bugs, it's a good practice to include a `return` here. If I were to something after this line, I would silently introduce a bug, because the code would get executed, even when `Delete()` returns an error:

```golang
if err != nil {
		*errMessage = C.CString(err.Error())
}
fmt.Println("All went well!") // Well yes, but actually no.
```
And best thing of all: it won't hurt the performance. The compiler will almost certainly not make a difference between there being a `return` or not.

```suggestion
		*errMessage = C.CString(err.Error())
		return
```

Reply:
Nice catch. Fixed.